### DISCUSSION TITLE
Passing props to page via navigateTo

### DISCUSSION THREAD

**Discussion by husayt** [nuxt/nuxt#30105] [Ideas]
It seems there’s currently no built-in way to pass props directly to a page during internal navigation in Nuxt, which feels like a missed opportunity. Since each Nuxt page is essentially a Vue component that can accept props, introducing a props parameter to `navigateTo` for passing properties to the target page-component would be incredibly useful.

Right now, the workaround is to URL-encode these props into query parameters and extract them on the page load, but this approach feels less elegant and more cumbersome. (other workarounds would be to use useState or etc)
.

If there’s already a solution or alternative approach for this that I’ve overlooked, please let me know. Otherwise, adding this capability could truly elevate navigateTo and make internal navigation much more powerful.

**Comment by TheAlexLichter** [comment]
When navigating to a new page, passing props beyond whatever is in the state of the URL is not a pattern that should be encouraged because it will change behavior between the initial render and navigating from another site and thus make it somewhat less predictable.

That this worked in Vue 2 was a side effect and not desired ([source](https://x.com/themarcba/status/1671798853601820673))

**Reply by husayt** [reply]
Hi Alex,
Thank you for the clarification. While I understand the need for caution, I believe this feature could be implemented in a balanced and safe way. The Nuxt 3 architecture is much more mature than 2 was, and already supports mechanisms to handle such scenarios (type safety and etc). Similar to standard Vue components, if no props are passed, they would simply default to empty values.

These props could remain entirely optional, applicable only for internal navigation. The navigateTo method could include an additional parameter specifically for passing props during internal routing.

In my case, I implemented a workaround using `useState`, but it feels somewhat hacky and not entirely in the spirit of Nuxt's design philosophy. A built-in solution would streamline the process and enhance developer experience.
